src/ContentEditableInput.js

```diff
@@ -0,0 +1,469 @@
+    if (toIndex == display.view.length - 1) {
+      var toLine = display.viewTo - 1;
+      var toNode = display.lineDiv.lastChild;
+    } else {
+      var toLine = lineNo(display.view[toIndex + 1].line) - 1;
+      var toNode = display.view[toIndex + 1].node.previousSibling;
+    }
+
+    var newText = cm.doc.splitLines(domTextBetween(cm, fromNode, toNode, fromLine, toLine));
+    var oldText = getBetween(cm.doc, Pos(fromLine, 0), Pos(toLine, getLine(cm.doc, toLine).text.length));
+    while (newText.length > 1 && oldText.length > 1) {
+      if (lst(newText) == lst(oldText)) { newText.pop(); oldText.pop(); toLine--; }
+      else if (newText[0] == oldText[0]) { newText.shift(); oldText.shift(); fromLine++; }
+      else break;
+    }
+
+    var cutFront = 0, cutEnd = 0;
+    var newTop = newText[0], oldTop = oldText[0], maxCutFront = Math.min(newTop.length, oldTop.length);
+    while (cutFront < maxCutFront && newTop.charCodeAt(cutFront) == oldTop.charCodeAt(cutFront))
+      ++cutFront;
+    var newBot = lst(newText), oldBot = lst(oldText);
+    var maxCutEnd = Math.min(newBot.length - (newText.length == 1 ? cutFront : 0),
+                             oldBot.length - (oldText.length == 1 ? cutFront : 0));
+    while (cutEnd < maxCutEnd &&
+           newBot.charCodeAt(newBot.length - cutEnd - 1) == oldBot.charCodeAt(oldBot.length - cutEnd - 1))
+      ++cutEnd;
+
+    newText[newText.length - 1] = newBot.slice(0, newBot.length - cutEnd);
+    newText[0] = newText[0].slice(cutFront);
+
+    var chFrom = Pos(fromLine, cutFront);
+    var chTo = Pos(toLine, oldText.length ? lst(oldText).length - cutEnd : 0);
+    if (newText.length > 1 || newText[0] || cmp(chFrom, chTo)) {
+      replaceRange(cm.doc, newText, chFrom, chTo, "+input");
+      return true;
+    }
+  },
+
+  ensurePolled: function() {
+    this.forceCompositionEnd();
+  },
+  reset: function() {
+    this.forceCompositionEnd();
+  },
+  forceCompositionEnd: function() {
+    if (!this.composing || this.composing.handled) return;
+    this.applyComposition(this.composing);
+    this.composing.handled = true;
+    this.div.blur();
+    this.div.focus();
+  },
+  applyComposition: function(composing) {
+    if (this.cm.isReadOnly())
+      operation(this.cm, regChange)(this.cm)
+    else if (composing.data && composing.data != composing.startData)
+      operation(this.cm, applyTextInput)(this.cm, composing.data, 0, composing.sel);
+  },
+
+  setUneditable: function(node) {
+    node.contentEditable = "false"
+  },
+
+  onKeyPress: function(e) {
+    e.preventDefault();
+    if (!this.cm.isReadOnly())
+      operation(this.cm, applyTextInput)(this.cm, String.fromCharCode(e.charCode == null ? e.keyCode : e.charCode), 0);
+  },
+
+  readOnlyChanged: function(val) {
+    this.div.contentEditable = String(val != "nocursor")
+  },
+
+  onContextMenu: nothing,
+  resetPosition: nothing,
+
+  needsContentAttribute: true
+  }, ContentEditableInput.prototype);
+
+function posToDOM(cm, pos) {
```

Comment:
I'd put all the scary DOM reading stuff that follows close to the display drawing code, rather than here (yeah, I know, it was here in the original file as well)